File: web/src/db/repositories/note.ts
/**
 * Note 和 Card 本地存储操作
 */

import { nanoid } from 'nanoid'
import { db } from '@/db'
import type { Note, Card, SourceType } from '@/types'

/**
 * 根据字段内容生成 GUID
 */
function generateGuid(fields: Record<string, string>): string {
  const content = Object.values(fields).find((v) => v) || ''
  // 简单哈希
  let hash = 0
  for (let i = 0; i < content.length; i++) {
    const char = content.charCodeAt(i)
    hash = (hash << 5) - hash + char
    hash = hash & hash
  }
  return Math.abs(hash).toString(36)
}

export const noteRepo = {
  /**
   * 获取牌组的所有笔记
   */
  async getByDeckId(deckId: string): Promise<Note[]> {
    return db.notes
      .where('deckId')
      .equals(deckId)
      .and((n) => !n.deletedAt)
      .toArray()
  },

  /**
   * 根据 ID 获取笔记（包含卡片）
   */
  async getById(id: string): Promise<(Note & { cards: Card[] }) | undefined> {
    const note = await db.notes.get(id)
    if (note && !note.deletedAt) {
      const cards = await db.cards
        .where('noteId')
        .equals(id)
        .and((c) => !c.deletedAt)
        .toArray()
      return { ...note, cards }
    }
    return undefined
  },

  /**
   * 创建笔记（同时创建卡片）
   */
  async create(
    userId: string,
    deckId: string,
    noteModelId: string,
    fields: Record<string, string>,
    templateIds: string[],
    options?: {
      tags?: string[]
      sourceType?: SourceType
      sourceMeta?: Record<string, unknown>
    }
  ): Promise<Note> {
    const now = Date.now()
    const noteId = nanoid()

    const note: Note = {
      id: noteId,
      userId,
      deckId,
      noteModelId,
      guid: generateGuid(fields),
      fields,
      tags: options?.tags || [],
      sourceType: options?.sourceType || 'manual',
      sourceMeta: options?.sourceMeta,
      dirty: 1,
      createdAt: now,
      updatedAt: now,
    }

    // 为每个模板创建卡片
    const cards: Card[] = templateIds.map((templateId, index) => ({
      id: nanoid(),
      userId,
      noteId,
      deckId,
      cardTemplateId: templateId,
      ord: index,
      state: 'new' as const,
      queue: 'new' as const,
      due: 0,
      interval: 0,
      easeFactor: 2500,
      reps: 0,
      lapses: 0,
      stability: 0,
      difficulty: 0,
      dirty: 1,
      createdAt: now,
      updatedAt: now,
    }))

    await db.transaction('rw', [db.notes, db.cards], async () => {
      await db.notes.add(note)
      await db.cards.bulkAdd(cards)
    })

    return note
  },

  /**
   * 更新笔记
   */
  async update(
    id: string,
    data: Partial<Pick<Note, 'fields' | 'tags' | 'deckId'>>
  ): Promise<void> {
    const updates: Partial<Note> = {
      ...data,
      dirty: 1,
      updatedAt: Date.now(),
    }
    if (data.fields) {
      updates.guid = generateGuid(data.fields)
    }
    await db.notes.update(id, updates)
  },

  /**
   * 删除笔记（软删除，同时删除关联卡片）
   */
  async delete(id: string): Promise<void> {
    const now = Date.now()
    await db.transaction('rw', [db.notes, db.cards], async () => {
      await db.notes.update(id, { deletedAt: now, updatedAt: now })
      await db.cards
        .where('noteId')
        .equals(id)
        .modify({ deletedAt: now, updatedAt: now })
    })
  },

  /**
   * 搜索笔记
   */
  async search(userId: string, keyword: string): Promise<Note[]> {
    const notes = await db.notes
      .where('userId')
      .equals(userId)
      .and((n) => !n.deletedAt)
      .toArray()

    const lowerKeyword = keyword.toLowerCase()
    return notes.filter((note) =>
      Object.values(note.fields).some((value) =>
        value.toLowerCase().includes(lowerKeyword)
      )
    )
  },
}

export const cardRepo = {
  /**
   * 获取牌组的所有卡片
   */
  async getByDeckId(deckId: string): Promise<Card[]> {
    return db.cards
      .where('deckId')
      .equals(deckId)
      .and((c) => !c.deletedAt)
      .toArray()
  },

  /**
   * 根据 ID 获取卡片
   */
  async getById(id: string): Promise<Card | undefined> {
    const card = await db.cards.get(id)
    return card && !card.deletedAt ? card : undefined
  },

  /**
   * 获取待复习的卡片
   */
  async getDueCards(
    userId: string,
    options?: {
      deckId?: string
      dueBefore?: number
      limit?: number
    }
  ): Promise<Card[]> {
    const now = options?.dueBefore ?? Date.now()
    let query = db.cards
      .where('userId')
      .equals(userId)
      .and((c) => !c.deletedAt && c.queue !== 'suspended' && c.due <= now)

    if (options?.deckId) {
      const cards = await query.toArray()
      const filtered = cards.filter((c) => c.deckId === options.deckId)
      return options?.limit ? filtered.slice(0, options.limit) : filtered
    }

    const cards = await query.toArray()
    return options?.limit ? cards.slice(0, options.limit) : cards
  },

  /**
   * 更新卡片（调度状态）
   */
  async update(id: string, data: Partial<Card>): Promise<void> {
    await db.cards.update(id, {
      ...data,
      dirty: 1,
      updatedAt: Date.now(),
    })
  },

  /**
   * 统计卡片状态
   */
  async countByState(userId: string, deckId?: string): Promise<Record<string, number>> {
    const cards = await db.cards
      .where('userId')
      .equals(userId)
      .and((c) => !c.deletedAt && (!deckId || c.deckId === deckId))
      .toArray()

    return cards.reduce(
      (acc, card) => {
        acc[card.state] = (acc[card.state] || 0) + 1
        return acc
      },
      {} as Record<string, number>
    )
  },
}

File: app/schemas/note.py
"""
笔记和卡片相关的 Pydantic Schema

用于 API 请求和响应的数据验证和序列化
"""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

# ==================== 卡片 Schema ====================


class CardBase(BaseModel):
    """卡片基础字段"""

    state: Literal["new", "learning", "review", "relearning"] = Field(default="new", description="状态")
    queue: Literal["new", "learning", "review", "suspended"] = Field(default="new", description="队列")
    due: int = Field(default=0, description="下次复习时间戳或天编号")
    interval: int = Field(default=0, ge=0, description="当前间隔（天）")
    ease_factor: int = Field(default=2500, ge=1300, description="难度系数（2500=2.5）")


class CardResponse(CardBase):
    """卡片响应"""

    id: str = Field(..., description="卡片ID")
    user_id: str = Field(..., description="所属用户ID")
    note_id: str = Field(..., description="所属笔记ID")
    deck_id: str = Field(..., description="所属牌组ID")
    card_template_id: str = Field(..., description="卡片模板ID")
    ord: int = Field(default=0, description="模板序号")
    reps: int = Field(default=0, description="复习次数")
    lapses: int = Field(default=0, description="遗忘次数")
    last_review: int | None = Field(default=None, description="上次复习时间戳")
    stability: float = Field(default=0.0, description="FSRS 稳定性")
    difficulty: float = Field(default=0.0, description="FSRS 难度")
    created_at: datetime | None = Field(default=None, description="创建时间")
    updated_at: datetime | None = Field(default=None, description="更新时间")

    model_config = {"from_attributes": True}


class CardUpdate(BaseModel):
    """更新卡片请求（主要用于调度状态更新）"""

    state: Literal["new", "learning", "review", "relearning"] | None = Field(default=None, description="状态")
    queue: Literal["new", "learning", "review", "suspended"] | None = Field(default=None, description="队列")
    due: int | None = Field(default=None, description="下次复习时间戳")
    interval: int | None = Field(default=None, ge=0, description="当前间隔（天）")
    ease_factor: int | None = Field(default=None, ge=1300, description="难度系数")
    reps: int | None = Field(default=None, ge=0, description="复习次数")
    lapses: int | None = Field(default=None, ge=0, description="遗忘次数")
    last_review: int | None = Field(default=None, description="上次复习时间戳")
    stability: float | None = Field(default=None, description="FSRS 稳定性")
    difficulty: float | None = Field(default=None, description="FSRS 难度")


# ==================== 笔记 Schema ====================


class NoteBase(BaseModel):
    """笔记基础字段"""

    deck_id: str = Field(..., description="所属牌组ID")
    note_model_id: str = Field(..., description="笔记类型ID")
    fields: dict[str, str] = Field(..., description="字段内容")
    tags: list[str] = Field(default_factory=list, description="标签列表")


class NoteCreate(NoteBase):
    """创建笔记请求"""

    source_type: Literal["manual", "ai", "import"] = Field(default="manual", description="来源类型")
    source_meta: dict | None = Field(default=None, description="来源元数据")


class NoteUpdate(BaseModel):
    """更新笔记请求"""

    deck_id: str | None = Field(default=None, description="所属牌组ID")
    fields: dict[str, str] | None = Field(default=None, description="字段内容")
    tags: list[str] | None = Field(default=None, description="标签列表")


class NoteResponse(NoteBase):
    """笔记响应"""

    id: str = Field(..., description="笔记ID")
    user_id: str = Field(..., description="所属用户ID")
    guid: str = Field(..., description="语义唯一标识")
    source_type: str = Field(default="manual", description="来源类型")
    source_meta: dict | None = Field(default=None, description="来源元数据")
    cards: list[CardResponse] = Field(default_factory=list, description="关联的卡片")
    created_at: datetime | None = Field(default=None, description="创建时间")
    updated_at: datetime | None = Field(default=None, description="更新时间")

    model_config = {"from_attributes": True}


class NoteListQuery(BaseModel):
    """笔记列表查询参数"""

    deck_id: str | None = Field(default=None, description="牌组ID")
    keyword: str | None = Field(default=None, description="搜索关键词（字段内容）")
    tags: list[str] | None = Field(default=None, description="标签过滤")


class CardListQuery(BaseModel):
    """卡片列表查询参数"""

    deck_id: str | None = Field(default=None, description="牌组ID")
    state: Literal["new", "learning", "review", "relearning"] | None = Field(default=None, description="状态过滤")
    queue: Literal["new", "learning", "review", "suspended"] | None = Field(default=None, description="队列过滤")
    due_before: int | None = Field(default=None, description="到期时间之前")
